streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -375,150 +199,117 @@ private RuntimeException unAssignChangeLogPartitions() {
-                log.error("{} Failed to {} " + taskType + " {} due to the following error:",
-                          logPrefix,
-                          action.name(),
-                          task.id(),
-                          t);
-                if (firstException == null) {
-                    firstException = t;
-                }
-            }
-        }
-
-        return firstException;
-    }
-
-
-
     void shutdown(final boolean clean) {
         log.debug("{} Shutting down all active tasks {}, standby tasks {}, suspended tasks {}, and suspended standby tasks {}",
-                  logPrefix, activeTasks.keySet(), standbyTasks.keySet(),
-                  suspendedTasks.keySet(), suspendedStandbyTasks.keySet());
-
-        for (final Task task : allTasks()) {
-            try {
-                task.close(clean);
-            } catch (final RuntimeException e) {
-                log.error("{} Failed while closing {} {} due to the following error:",
-                          logPrefix,
-                          task.getClass().getSimpleName(),
-                          task.id(),
-                          e);
-            }
-        }
+                  logPrefix, active.runningTaskIds(), standby.runningTaskIds(),
+                  active.previousTaskIds(), standby.previousTaskIds());
+
+        active.close(clean);
+        standby.close(clean);
         try {
             threadMetadataProvider.close();
         } catch (final Throwable e) {
             log.error("{} Failed to close KafkaStreamClient due to the following error:", logPrefix, e);
         }
         // remove the changelog partitions from restore consumer
         unAssignChangeLogPartitions();
+        taskCreator.close();
 
     }
 
     Set<TaskId> suspendedActiveTaskIds() {
-        return Collections.unmodifiableSet(suspendedTasks.keySet());
+        return active.previousTaskIds();
     }
 
     Set<TaskId> suspendedStandbyTaskIds() {
-        return Collections.unmodifiableSet(suspendedStandbyTasks.keySet());
+        return standby.previousTaskIds();
     }
 
-    void removeTasks() {
-        removeStreamTasks();
-        removeStandbyTasks();
+    Task activeTask(final TopicPartition partition) {
+        return active.runningTaskFor(partition);
     }
 
-    private void removeStandbyTasks() {
-        log.debug("{} Removing all standby tasks {}", logPrefix, standbyTasks.keySet());
-        standbyTasks.clear();
-        standbyTasksByPartition.clear();
+
+    Task standbyTask(final TopicPartition partition) {
+        return standby.runningTaskFor(partition);
     }
 
-    Task activeTask(final TopicPartition partition) {
-        return activeTasksByPartition.get(partition);
+    Map<TaskId, Task> activeTasks() {
+        return active.runningTaskMap();
     }
 
-    boolean hasStandbyTasks() {
-        return !standbyTasks.isEmpty();
+    void setConsumer(final Consumer<byte[], byte[]> consumer) {
+        this.consumer = consumer;
```

Comment:
Personally I'd suggest remove setters if it is not called multiple times with different parameters within its life cycle, unless we cannot work around it.

For this case we can create the consumer in `public static StreamThread create()` and pass it into the TaskAssignor and StreamThread constructors and to remove this setter function.